stl/inc/format

```diff
@@ -457,18 +457,34 @@ _NODISCARD constexpr const _CharT* _Parse_arg_id(
     _THROW(format_error("Invalid format string."));
 }
 
-_NODISCARD constexpr bool _Is_execution_charset_utf8() {
-#pragma warning(push)
-#pragma warning(disable : 4309) // 'initializing' : truncation of constant value
-#pragma warning(disable : 4566) // character represented by universal-character-name '\u4E00' cannot be represented in
-                                // the current code page
-#pragma warning(disable : 6201) // Index '2' is out of valid index range '0' to '1' for possibly stack allocated buffer
-                                // '_Test_char'
-#pragma warning(disable : 6239) // (<non-zero constant> && <expression>) always evaluates to the result of <expression>.
-                                // Did you intend to use the bitwise-and operator?
-    constexpr char _Test_char[] = "\u4e00";
-    return sizeof(_Test_char) == 4 && _Test_char[0] == '\xe4' && _Test_char[1] == '\xb8' && _Test_char[2] == '\x80';
-#pragma warning(pop)
+_NODISCARD constexpr bool _Is_execution_charset_self_synchronizing() {
+#ifdef __clang__ // TRANSITION, LLVM-52549 (clang-cl 14 will define _MSVC_EXECUTION_CHARACTER_SET)
+    return string_view{__clang_literal_encoding__} == "UTF-8";
+#elif defined(__EDG__) // TRANSITION, VSO-1468747
+    return false; // safe fallback response
+#else // ^^^ EDG / MSVC vvv
+    // This is a list of charset codes (as reported by _MSVC_EXECUTION_CHARACTER_SET) that are
+    // self-synchronizing (and "narrow" so 1200 (utf-16) and 1201 (unicodeFFFE) are not listed).
+    // We care about this because if a charset is self-synchronizing then we can search through it
+    // for formatting control characters _without_ generally decoding the format string.
+    switch (_MSVC_EXECUTION_CHARACTER_SET) {

```

Comment:
Shouldn't `#ifdef _MSVC_EXECUTION_CHARACTER_SET` be the first branch, so new compilers that learn to define the macro will light up the correct branch without requiring us to touch the source? `#ifdef _M_E_C_S #elif d(__clang__) #else` seems better to me than assuming clang/EDG don't define the macro. (I test Clang 14 from time to time.)

Reply:
I generally worry about automatically lighting up code (since it can go wrong if the updated compiler doesn't behave exactly as we expect), although the risk here seems low. I can make this change.